### DISCUSSION TITLE
Coding Style

### DISCUSSION THREAD

**Discussion by DanielTOsborne** [opendcs/opendcs#50] [Development Practices]
With a large project such as this, with many external developers, it's important to have a defined coding style/guide.

In addition, we should set up tools to help users stay within the style guide.
This can include items such as a .editorconfig, or linting.

As I'm sure most developers have experienced at some point, I personally don't like all the existing coding style used. However as long as the style is documented, I'll write code to fit the existing style.

**Comment by mikecovesw** [comment]
You're going to start putting open-brackets at the ends of lines, aren't 
you? :-)


On 4/29/22 11:18 AM, Daniel Osborne wrote:
>
> With a large project such as this, with many external developers, it's 
> important to have a defined coding style/guide.
>
> In addition, we should set up tools to help users stay within the 
> style guide.
> This can include items such as a .editorconfig, or linting.
>
> As I'm sure most developers have experienced at some point, I 
> personally don't like all the existing coding style used. However as 
> long as the style is documented, I'll write code to fit the existing 
> style.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/opendcs/opendcs/discussions/50>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABTDKHSFMWV3MZVJLJAZODDVHP4WFANCNFSM5UWLZU2Q>.
> You are receiving this because you are subscribed to this 
> thread.Message ID: ***@***.***>
>
-- 
Michael Maloney
Cove Software, LLC
Tel: (410) 241-5248
***@***.***
Web:www.covesw.com

**Reply by DanielTOsborne** [reply]
Ha! No.
That I actually can't stand. I know it's common in Java, but I don't like that.

Not to call you out, but....
Mixed tabs and spaces is a big pet peeve of mine. 😄 

https://github.com/opendcs/opendcs/blob/master/src/main/java/decodes/cwms/HecConstants.java

**Comment by MikeNeilson** [comment]
Worth asking the group but I feel like using the next line is the right choice because of how much is already that way.

Daniel perhaps you can propose the the initial form and we can comment on it. 

editor config is good, checkstyle that we run manually and on pullrequest would probably be good as well.

**Reply by DanielTOsborne** [reply]
I think starting with how the code is already structured is the best course of action, since for the most part, the code up to now has had very few developers working on it.

**Comment by MikeNeilson** [comment]
So far we've got

- { on the next line
- one type of whitespace. should we do tabs or spaces? seems like it's already spaces but maybe a sanity check
-

**Comment by agilmore2** [comment]
spaces. Tabs for whitespace are so last century. ;)

**Comment by mikecovesw** [comment]
I don't care as long as it's consistent. I normally set the editor use 
tabs and set tabstop to 4. What messes me up is folks who "assume" tabs 
are worth 8 spaces and mix a tab and 4 spaces to mean 3 indents.

On 5/4/22 12:03 AM, agilmore2 wrote:
> spaces. Tabs for whitespace are so last century. ;)
>
> On Tue, May 3, 2022 at 11:56 AM Mike Neilson ***@***.***>
> wrote:
>
> > So far we've got
> >
> > - { on the next line
> > - one type of whitespace. should we do tabs or spaces? seems like it's
> > already spaces but maybe a sanity check
> > -
> >
> > —
> > Reply to this email directly, view it on GitHub
> > 
> <https://github.com/opendcs/opendcs/discussions/50#discussioncomment-2682075>,
> > or unsubscribe
> > 
> <https://github.com/notifications/unsubscribe-auth/ABUI5SUMKZX6IPW7HLZD3U3VIFSF5ANCNFSM5UWLZU2Q>
> > .
> > You are receiving this because you are subscribed to this thread.Message
> > ID: ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/opendcs/opendcs/discussions/50#discussioncomment-2684287>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABTDKHTMOAB43LT2BYXG7ATVIHZJVANCNFSM5UWLZU2Q>.
> You are receiving this because you commented.Message ID: 
> ***@***.***>
>
-- 
Michael Maloney
Cove Software, LLC
Tel: (410) 241-5248
***@***.***
Web:www.covesw.com

**Comment by MikeNeilson** [comment]
I've got the start of a checkstyle file in https://github.com/opendcs/opendcs/pull/74

It's not part of the primary workflow. I only have some basic elements to show what's doable we need to start filling out the rest of it and decide if we should just reformat the code base in one go (later, part of that PR).

There are github build actions that will report *only* on the code that a PR submitter changed. I think that's a good place to start. But it would be good to just decide on a style and just apply it. There's quite the combination of tabs,spaces and which line the {'s start on.

**Reply by DanielTOsborne** [reply]
Assuming the majority of the existing style doesn't diverge from the recommended Java style, I'd say we start there.
It's located here (https://www.oracle.com/java/technologies/javase/codeconventions-introduction.html), but I've noticed the formatting of examples on the website is erroneous, probably due to how old it is.
If we can find a PDF, or older (pre-Oracle) version archive, that might display better.

**Comment by MikeNeilson** [comment]
I think it mostly is. There's a places the old sun checkstyle file exists. I'll dig it up and adjust for the things we discussed above.

**Comment by MikeNeilson** [comment]
Style has been agreed upon and we are starting to implement. Only thing we really lack is auto formatting files within the project. However, that is IDE specific. But we may still implement a few.
